### PR TITLE
upgrade PTL trainer flags

### PR DESCRIPTION
**Pull request by nithinraok**
Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

# What does this PR do ?

Upgrade pytorch_lightning trainer flags that are going to be deprecated in future

**Collection**: Core and All collections yaml files 

# Changelog 
- Update pytorch_lightning deprecated variables, namely:

> -  accelerator to the trainer now takes the type of accelerator device than distributed type
> -  distributed type is handled by strategy flag
> -  previously: accelerator='ddp' now accelerator='gpu' strategy='ddp'
> -  trainer flag checkpoint_callback is renamed to enable_checkpointing
> - ModelCheckpoint argument evert_n_val_epochs renamed to every_n_epochs
> - Update `trainer.gpus` to `trainer.devices` ( With this change setting devices=0 doesnt auto select gpu instead need to make accelerator='cpu' and also note accelerator=cpu now will only accept >0 as values for devices) 
> - `examples/asr/transcribe_speech.py` script modified to remove cuda option. Script automatically searches for GPU and if its not present then falls back to CPU  
> - `checkpointing_callback` -> `enable_checkpointing` 
> - `flush_logs_every_n_steps` removed from trainer flags
> - `sync_batchnorm` and `num_nodes` removed from NLPDPPPlugins 

# Usage
- 

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [] Did you write any new necessary tests?
- [] Did you add or update any necessary documentation?
- [] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
-

### CODE REVIEW

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 1]
```diff
@@ -65,12 +63,12 @@ def __init__(
         no_ddp_communication_hook: bool = False,
         **kwargs: Union[Any, Dict[str, Any]],
     ) -> None:
+        super().__init__(parallel_devices, num_nodes, cluster_environment, checkpoint_io, sync_batchnorm, **kwargs)
```
Comment (titu1994) [inline nemo/collections/nlp/parts/nlp_overrides.py:66]: Huh ?? Why these changes ?

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 2]
```diff
@@ -65,12 +63,12 @@ def __init__(
...
+
         if not HAVE_APEX:
-            raise ImportError(
+            logging.warning(
```
Comment (titu1994) [inline nemo/collections/nlp/parts/nlp_overrides.py:69]: Why revert this ?

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 3]
```diff
@@ -203,11 +201,11 @@ def distributed_sampler_kwargs(self):
 
 class NLPSaveRestoreConnector(SaveRestoreConnector):
     def __init__(self) -> None:
+        super().__init__()
```
Comment (titu1994) [inline nemo/collections/nlp/parts/nlp_overrides.py:204]: Why change these ?

File: nemo/collections/tts/helpers/helpers.py [hunk 4]
```diff
@@ -91,15 +91,16 @@ def get_batch_size(train_dataloader):
 
 
 def get_num_workers(trainer):
-    if trainer.accelerator is None:
+    if trainer._distrib_type is None:
```
Comment (titu1994) [inline nemo/collections/tts/helpers/helpers.py:94]: There's no public API for these anymore ? Can you make sure to use public API if possible ? FYI @blisc

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -473,18 +473,18 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
                 optim_config['sched']['t_max_epochs'] = self._trainer.max_epochs
                 optim_config['sched']['t_accumulate_grad_batches'] = self._trainer.accumulate_grad_batches
                 optim_config['sched']['t_limit_train_batches'] = self._trainer.limit_train_batches
-                if self._trainer.accelerator is None:
+                if self._trainer._distrib_type is None:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:476]: Same as above.

File: requirements/requirements_lightning.txt [hunk 6]
```diff
@@ -4,4 +4,4 @@ transformers>=4.0.1
 webdataset>=0.1.48,<=0.1.62
 omegaconf>=2.1.0
 hydra-core>=1.1.0
-pyyaml<6  # Pinned until omegaconf works with pyyaml>=6
\ No newline at end of file
+pyyaml<6  # Pinned until omegaconf works with pyyaml>=6
```
Comment (titu1994) [inline requirements/requirements_lightning.txt:7]: ?

**Review by titu1994 — COMMENTED** [review]
The day of 100 file review is here :). 

Most of it looks great, amazing work here. 
Some comments about why nlp stuff was changed and whether we can use public API to test for accelerator type

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 1 above)
Comment (nithinraok) [reply nemo/collections/nlp/parts/nlp_overrides.py]: huh great catch, I dont know how these changes crept in. Probably when I rebased from r1.6.0 to main. Removed it :)

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/nlp/parts/nlp_overrides.py]: reverted. Not an intentional change

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 3 above)
Comment (nithinraok) [reply nemo/collections/nlp/parts/nlp_overrides.py]: reverted

File: requirements/requirements_lightning.txt (hunk 6 above)
Comment (nithinraok) [reply requirements/requirements_lightning.txt:7]: there is no change in that line?

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/tts/helpers/helpers.py:94]: Its not clear from their APIs but this is how I found about `distrib_type`. Rest others like `strategy_type` `distributed_backend` aren't always set for various trainer initialization cases

Comment (titu1994) [reply nemo/collections/tts/helpers/helpers.py:94]: That's a PL bug then if it's not set correctly always. We should reach out to them about this, using private variable for a critical part just means it will silently break.

File: requirements/requirements_lightning.txt (hunk 6 above)
Comment (titu1994) [reply requirements/requirements_lightning.txt:7]: Huh ok

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/nlp/parts/nlp_overrides.py]: Ah lol ok.

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/nlp/parts/nlp_overrides.py]: Hehe

**Review by blisc — CHANGES_REQUESTED** [review]

File: examples/asr/asr_ctc/speech_to_text_ctc_bpe.py [hunk 7]
```diff
@@ -39,7 +39,8 @@
     model.tokenizer.dir=<path to directory of tokenizer (not full path to the vocab file!)> \
     model.tokenizer.type=<either bpe or wpe> \
     trainer.gpus=-1 \
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/asr/asr_ctc/speech_to_text_ctc_bpe.py:41]: Done.

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/tts/helpers/helpers.py:94]: This is weird to me. Could you also confirm the same?

There are some accelerator variables assigned to trainer properties, but I couldn't access them. 

```
from pytorch_lightning import Trainer
trainer=Trainer(devices=1,accelerator='gpu',strategy='dp')
print(trainer.num_gpus)# -> works
print(trainer.strategy) # doesnt work
```

File: nemo/utils/exp_manager.py [hunk 8]
```diff
@@ -75,7 +75,7 @@ class CallbackParams:
     save_top_k: Optional[int] = 3
     save_weights_only: Optional[bool] = False
     mode: Optional[str] = "min"
-    every_n_val_epochs: Optional[int] = 1
+    every_n_epochs: Optional[int] = 1
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/utils/exp_manager.py:78]: Thanks, good catch. Missed those.

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/tts/helpers/helpers.py:94]: @blisc my bad, I was viewing the master code base but had 1.5.9 running in the env. 

So currently this is in transition and I see no public api suggesting which property to use. 
I found `trainer.training_type_plugin.distributed_backend.value` would help us transition to 1.6 without any issues but that going to have a deprecated note from 1.6 and would be removed in 1.8 (as seen from master code). I can adobt this put a #TODO note so we can come back to this later once 1.6 is released and we make it a min version in our requirements

**Comment by nithinraok** [comment]
@titu1994 @blisc other than getting distributed type I made all the possible changes required for us to transition to PTL 1.6 . Let me know if we are missing anything. I changed the log to update all the changes I made, use it for reviewing :)

File: docs/source/asr/configs.rst [hunk 9]
```diff
@@ -752,7 +753,8 @@ Fine-tuning via a NeMo pretrained model name
         --config-name=<name of config without .yaml>) \
         model.train_ds.manifest_filepath="<path to manifest file>" \
         model.validation_ds.manifest_filepath="<path to manifest file>" \
-        trainer.gpus=-1 \
+        trainer.devices=-1 \
+        trainer.accelerator='gpu' \
```
Comment (titu1994) [inline docs/source/asr/configs.rst:757]: In the docs, there seems to be additional new line, upto you if you want to fix those (I would recommend to do it but there's a lot of files). If you can findreplace then try that c

File: examples/asr/asr_ctc/speech_to_text_ctc_bpe.py [hunk 10]
```diff
@@ -38,8 +38,9 @@
     model.validation_ds.manifest_filepath=<path to val/test manifest> \
     model.tokenizer.dir=<path to directory of tokenizer (not full path to the vocab file!)> \
     model.tokenizer.type=<either bpe or wpe> \
-    trainer.gpus=-1 \
-    trainer.accelerator="ddp" \
+    trainer.devices=-1 \
+    trainer.accelerator="gpu" \
+    strategy="ddp" \
```
Comment (titu1994) [inline examples/asr/asr_ctc/speech_to_text_ctc_bpe.py:43]: Strategy seperate from trainer ?

File: examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py [hunk 11]
```diff
@@ -38,8 +38,9 @@
     model.validation_ds.manifest_filepath=<path to val/test manifest> \
     model.tokenizer.dir=<path to directory of tokenizer (not full path to the vocab file!)> \
     model.tokenizer.type=<either bpe or wpe> \
-    trainer.gpus=-1 \
-    trainer.accelerator="ddp" \
+    trainer.devices=-1 \
+    trainer.accelerator="gpu" \
+    strategy="ddp" \
```
Comment (titu1994) [inline examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py:43]: SSame As above

File: examples/asr/transcribe_speech.py [hunk 12]
```diff
@@ -61,7 +60,6 @@
     dataset_manifest="" \
     output_filename="" \
     batch_size=32 \
-    cuda=0 \
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:64]: Don't remove this. Put it back as int since we need it to select the GPU at runtime for multi gpu transcribe.

File: examples/asr/transcribe_speech.py [hunk 13]
```diff
@@ -79,10 +77,6 @@ class TranscriptionConfig:
     batch_size: int = 32
     num_workers: int = min(batch_size, os.cpu_count() - 1)
 
-    # Set `cuda` to int to define CUDA device. If 'None', will look for CUDA
-    # device anyway, and do inference on CPU only if CUDA device is not found.
-    # If `cuda` is a negative number, inference will be on CPU only.
-    cuda: Optional[int] = None
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:85]: Revert c

File: examples/asr/transcribe_speech.py [hunk 14]
```diff
@@ -103,13 +97,14 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
         raise ValueError("Both cfg.audio_dir and cfg.dataset_manifest cannot be None!")
 
     # setup GPU
-    if cfg.cuda is None:
-        if torch.cuda.is_available():
-            cfg.cuda = 0  # use 0th CUDA device
-        else:
-            cfg.cuda = -1  # use CPU
+    if torch.cuda.is_available():
+        accelerator = 'gpu'
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:101]: Don't select 0th GPU - let user decide c

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/tts/helpers/helpers.py:94]: I support that for now - add a to-do to all places where we need such values from trainer for easy find later, and let's use this deprecated arg for now until Nemo 1.7 is release. We need to discuss with them to have a replacement for critical arguments.

**Review by titu1994 — APPROVED** [review]
160+ files and counting. I commend you @nithinraok for managing to update this PR. I would give up a long time ago 🤣 

Mostly looks good, one issues with transcribe_speech but that's easy to resolve.

File: docs/source/asr/configs.rst (hunk 9 above)
Comment (nithinraok) [reply docs/source/asr/configs.rst:757]: Probably we can look this next time, too many files already

File: examples/asr/asr_ctc/speech_to_text_ctc_bpe.py (hunk 10 above)
Comment (nithinraok) [reply examples/asr/asr_ctc/speech_to_text_ctc_bpe.py:43]: Good catch :)

File: examples/asr/transcribe_speech.py (hunk 12 above)
Comment (nithinraok) [reply examples/asr/transcribe_speech.py:64]: I actually recommend this and user can choose the cuda device with "CUDA_VISIBLE_DEVICES=DEVICE_ID"

File: examples/asr/transcribe_speech.py (hunk 13 above)
Comment (nithinraok) [reply examples/asr/transcribe_speech.py:85]: Commented above

File: examples/asr/transcribe_speech.py (hunk 12 above)
Comment (titu1994) [reply examples/asr/transcribe_speech.py:64]: No it was decided that this remain an int. There's no need to change this.

Comment (nithinraok) [reply examples/asr/transcribe_speech.py:64]: Could you help me understand the reason? Because we have transcribe_parallel.py for multi-gpu setting anyway

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (blisc) [reply nemo/collections/tts/helpers/helpers.py]: Why do we need `strategy`? Can't we just return `devices * num_nodes` or something like that?

Comment (nithinraok) [reply nemo/collections/tts/helpers/helpers.py]: Unfortunetly this fails for some cases, when user passes `devices=-1` or `devices=[0,1..]` in this cases num_gpus would be good option

File: examples/asr/transcribe_speech.py (hunk 12 above)
Comment (titu1994) [reply examples/asr/transcribe_speech.py:64]: No this is the main script and it's API was set by the ASR team so it needs to remain the same as before.

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/tts/helpers/helpers.py]: Oh no, -1 is used across all the Slurm scripts ..

File: examples/asr/transcribe_speech.py (hunk 12 above)
Comment (nithinraok) [reply examples/asr/transcribe_speech.py:64]: Ok, will revert

File: nemo/collections/tts/helpers/helpers.py (hunk 4 above)
Comment (blisc) [reply nemo/collections/tts/helpers/helpers.py]: Ok, let's rip out this function. It is only used in univnet and hifigan where it is used in computing the max_steps for getting the learning rate. Don't we have some way of doing this in nemo core? Can we just replace this function in TTS with that one?
@titu1994 Do you remember where or if we have such a function?
@Oktai15 Can you confirm that this replace once found is appropriate here?

Comment (nithinraok) [reply nemo/collections/tts/helpers/helpers.py]: nemo core (modelPT.py) and tts helper (helpers.py) use same way to get num workers. I tried the latest PTL properties but nothing working as expected in a reasonable way (`strategy`, `devices*num_nodes`, `training_type_plugin`) . Its better to look at this after PTL 1.6 gets released or any minor version which brings the suitable property to upgrade to PTL 1.6 safely.

Comment (titu1994) [reply nemo/collections/tts/helpers/helpers.py]: It's in the setup_optimization() and I believe it's the same or similar to what's being done here. So all of our optimization would break too. @nithinraok have you reached out to PTL team regarding this ? What was their response if any ?

Comment (Oktai15) [reply nemo/collections/tts/helpers/helpers.py]: > Can you confirm that this replace once found is appropriate here?

@blisc, yes I can confirm that I took this code from modelPT.py

**Comment by nithinraok** [comment]
> @nithinraok, could you update the following too:
> 
> * UnivNet config https://github.com/NVIDIA/NeMo/blob/main/examples/tts/univnet.py
> * maybe remove `flush_logs_every_n_steps` from all configs, because it will be removed from PTL 1.7?

@Oktai15 Done

**Review by titu1994 — APPROVED** [review]
Lgtm, thanks for the enormous effort

**Review by ericharper — APPROVED** [review]
LGTM. Thank you!
